### PR TITLE
[`tests`] Extend training tests to include half-precision (bf16) and automatic mixed precision training

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

As per title and to specifically address: https://github.com/huggingface/transformers/pull/27220#pullrequestreview-1709756019

This PR aims to extend the current testing suite to include pure bfloat16 training (I think it is not possible to perform pure float16 training) + automatic mixed precision training through the `torch.amp` API: https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html 

That way we can easily check which model supports mixed precision training and support that from day 0 of the model integration

cc @amyeroberts @ydshieh 

Draft for now as many tests are failing

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -578,10 +579,20 @@ def check_training_gradient_checkpointing(self, gradient_checkpointing_kwargs=No
...
 
+        mixed_precision_dtype = torch.float16 if torch_device == "cuda" else torch.bfloat16
+        training_context = (
+            torch.amp.autocast(device_type=torch_device, dtype=mixed_precision_dtype)
+            if mixed_precision_training
+            else nullcontext()
+        )
+
+        def set_forward_fired(module, input, output):
+            module._hf_test_forward_fired = True
```
Comment (ydshieh) [inline tests/test_modeling_common.py:594]: Don't really think adding this dynamically is a good idea

**Comment by younesbelkada** [comment]
I did not had time to properly finish this PR, right now the blocker is that it appears for some architectures that some gradients are set to `None` when using GC even if I make sure that the forward pass has been correctly called. Except skipping the test for these archs, I am not sure how we should proceed here
